core/common/src/files/FileSystem.kt

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2010-2023 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE.txt file.
+ */
+
+package kotlinx.io.files
+
+import kotlinx.io.IOException
+import kotlinx.io.Sink
+import kotlinx.io.Source
+
+/**
+ * An interface providing basic operations on a file system.
+ *
+ * **This API is unstable and subject to change.**
+ */
+public sealed interface FileSystem {
```

Comment:
Could you please elaborate on why this is an interface and not a class?

Reply:
There should be a way to implement own filesystems for testing, at least. Of course, now it's impossible due to `sealed`, but at some point, we can open it.
There's not so much to reuse between different implementations, so I decided to abstain from making it an abstract class.